docs/en/reference/architecture.rst

```diff
@@ -1,25 +1,32 @@
 Architecture
 ============
 
-The DBAL consists of two layers: drivers and a wrapper. Each layer
-is mainly defined in terms of 3 components: ``Connection``,
+The DBAL consists of three layers: drivers, middlewares and a wrapper. Each
```

Comment:
I'd argue that middlewares are insignificant from the layering perspective. This is a way to conveniently extend the driver layer. The DBAL cannot work without the driver or the wrapper but it can work without using a middleware.

Reply:
I documented them as a third layer because as far as I understood them, they neither belong to the wrapper nor the driver layer. They sit in the middle of the two layers. I realize that an empty middleware layer is a perfectly valid state.

I can certainly reword this paragraph if have a better idea. But I felt that the middleware concept should appear on the architecture overview.

Reply:
> I documented them as a third layer because as far as I understood them, they neither belong to the wrapper nor the driver layer. They sit in the middle of the two layers. I realize that an empty middleware layer is a perfectly valid state.

I see it differently. A middleware is a way to extend the driver layer. I.e. it's more related to developer experience rather than the runtime API. All extensions made to the driver layer via a middleware could be done without one as well.

> I can certainly reword this paragraph if have a better idea. But I felt that the middleware concept should appear on the architecture overview.

My only ask is to revert the change related to the number of layers adding middleware as a layer. Otherwise, documenting middleware as an API is of course welcomed.

Reply:
Makes sense. I've reverted that part and adjusted the other paragraphs. Can you have a look again?